### PR TITLE
feat(core/tables, core/figures): support linked labels

### LINKED ISSUE #4222: Table and figure anchor link generation
Currently, figures and tables can be linked by id. For example, `<a href="#mytable"></a>`. This will replace the link text with the table number, such as "Table 3".

Linking to sections can be done in the other direction, however. If we have `<section><h2>My section</h2></section>` we can link to it by `<a>My section</a>`. Rather than link to the id, we can link to the name. Tables and figures could use the caption & figcaption for this.

Should ReSpec allow linking to tables & figures the same way it can link to sections?

### PR DESCRIPTION
**Pull request by relu91**
Hello 👋! this is my first PR here, I did my best to follow the [developer guide](https://github.com/w3c/respec/wiki/Developers-Guide) but feel free to point me to some wrongdoings. 

As the title describes, this PR converts the numbered labels of Tables and Figures from plain text to a *self-link* as asked in https://github.com/w3c/respec/issues/4222. Tests were updated accordingly but I didn't introduce any new tests specifically for this feature. Let me know if you feel like they should be added. Also, I tried to look around if there was some documentation to update but I wasn't sure how and where to add it. 

Closes #4222

### CODE REVIEW

**Comment by relu91** [comment]
**Note**: CI was already failing, this PR should not introduce any new falling tests.

**Comment by sidvishnoi** [comment]
I'll try to look into failing CI and review this PR by weekend.

**Comment by relu91** [comment]
> I'll try to look into failing CI and review this PR by weekend.

Thank you! let me know if I can be of any help :)

**Comment by relu91** [comment]
@sidvishnoi any news on the failing tests?

**Comment by sidvishnoi** [comment]
@relu91 Apologies, haven't got an opportunity to look into that yet. I'll try by this weekend, if not, we'll pretend the tests aren't failing.

File: src/core/anchor-expander.js [hunk 1]
```diff
@@ -82,13 +82,17 @@ function processFigure(matchingElement, id, a) {
...
     // @ts-ignore
-    node => !node.classList || !node.classList.contains("fig-title")
-  );
-  // drop an empty space at the end.
-  children.pop();
+    if (node.classList && node.classList.contains("figno")) {
+      // @ts-ignore
+      node.classList.remove("figno");
+    }
+    return node;
```
Comment (sidvishnoi) [inline src/core/anchor-expander.js:94]: Nit:
```suggestion
    node.classList?.remove("figno");
    return node;
```

**Review by sidvishnoi — APPROVED** [review]
PR looks good. Having tests would be nice, so we don't regress in future.

**Comment by relu91** [comment]
> PR looks good. Having tests would be nice, so we don't regress in future.

Added two minor tests they should help to keep track of breaking changes. Let me know if I can improve them.

File: src/core/anchor-expander.js [hunk 2]
```diff
@@ -109,13 +110,18 @@ function processTable(matchingTable, id, a) {
...
+  ].map(node => {
     // @ts-ignore
-    node => !node.classList || !node.classList.contains("table-title")
-  );
-  // drop an empty space at the end.
-  children.pop();
+    if (node.classList && node.classList.contains("tableno")) {
+      // @ts-ignore
+      node.classList.remove("tableno");
+    }
```
Comment (sidvishnoi) [inline src/core/anchor-expander.js:122]:
```suggestion
    // @ts-ignore
    node.classList?.remove("tableno");
```

File: tests/spec/core/figures-spec.js [hunk 3]
```diff
@@ -36,6 +36,22 @@ describe("Core - Figures", () => {
...
+       </figure>
+    `;
+    const ops = makeStandardOps(null, body);
+    const doc = await makeRSDoc(ops);
+
+    const caption = doc.getElementsByTagName("figcaption")[0];
+    const link = caption.querySelector("a");
+
+    expect(link.classList).toContain("self-link");
+    expect(link.href).toMatch(/#fig$/);
```
Comment (sidvishnoi) [inline tests/spec/core/figures-spec.js:52]:
```suggestion
    expect(link.href).toBe("#fig");
```

File: tests/spec/core/tables-spec.js [hunk 4]
```diff
@@ -30,6 +30,22 @@ describe("Core - Tables", () => {
...
+       </table>
+    `;
+    const ops = makeStandardOps(null, body);
+    const doc = await makeRSDoc(ops);
+
+    const caption = doc.getElementsByTagName("caption")[0];
+    const link = caption.querySelector("a");
+
+    expect(link.classList).toContain("self-link");
+    expect(link.href).toMatch(/#tab$/);
```
Comment (sidvishnoi) [inline tests/spec/core/tables-spec.js:46]:
```suggestion
    expect(link.href).toBe("#tab");
```

**Review by sidvishnoi — APPROVED** [review]
Thanks @relu91!
Do you have some comments on the dev process and guide that can be improved?

File: tests/spec/core/figures-spec.js [hunk 5]
```diff
@@ -36,6 +36,22 @@ describe("Core - Figures", () => {
...
+       </figure>
+    `;
+    const ops = makeStandardOps(null, body);
+    const doc = await makeRSDoc(ops);
+
+    const caption = doc.getElementsByTagName("figcaption")[0];
+    const link = caption.querySelector("a");
+
+    expect(link.classList).toContain("self-link");
+    expect(link.href).toBe("#fig");
```
Comment (sidvishnoi) [inline tests/spec/core/figures-spec.js:52]:
```suggestion
    expect(link.hash).toBe("#fig");
```

File: tests/spec/core/tables-spec.js [hunk 6]
```diff
@@ -30,6 +30,22 @@ describe("Core - Tables", () => {
...
+       </table>
+    `;
+    const ops = makeStandardOps(null, body);
+    const doc = await makeRSDoc(ops);
+
+    const caption = doc.getElementsByTagName("caption")[0];
+    const link = caption.querySelector("a");
+
+    expect(link.classList).toContain("self-link");
+    expect(link.href).toBe("#tab");
```
Comment (sidvishnoi) [inline tests/spec/core/tables-spec.js:46]:
```suggestion
    expect(link.hash).toBe("#tab");
```

**Comment by relu91** [comment]
> Do you have some comments on the dev process and guide that can be improved?

Not really, the development process was very smooth and it had a very easy onboarding and setup phase. Tests are pretty slow to run even if I have a reasonable performant station, but I guess you cannot do much about it. Thanks for fixing the tests, I didn't know about `hash` property, I should had pay more attention.

**Comment by sidvishnoi** [comment]
Good to know!

> Tests are pretty slow to run even if I have a reasonable performant station, but I guess you cannot do much about it

Yeah that's pain point. I believe it's due to reliance on network mostly. There are some plans (like using local resources in test build), but will have to find time for that.

Some new unrelated tests are broken right now, should be fixed automatically I think (else I'll merge https://github.com/w3c/respec/pull/4402 first). Will merge and release this PR today. 

Thanks again :)

**Comment by sidvishnoi** [comment]
BTW you can use `fit`, `fdescribe` (instead of `it()` and `describe()`) to run particular tests in dev.

**Comment by sidvishnoi** [comment]
Released in [`v32.8.0`](https://github.com/w3c/respec/releases/tag/v32.8.0) 🎉

**Comment by relu91** [comment]
Thank you really much!

**Pull request merged**